cpp/src/io/orc/orc.hpp

```diff
@@ -528,20 +532,24 @@ class ProtobufWriter {
                            TypeKind kind,
                            ColStatsBlob const* stats);
 
+  void resize(std::size_t bytes) { m_buf->resize(bytes); }
+
+  std::unique_ptr<std::vector<uint8_t>> release() { return std::move(m_buf); }
```

Comment:
looks like the `ProtobufWriter` object is not safe to use after `release()`. If possible, we should limit the `release()` use to rvalues.

Reply:
ideally, release should be called when `ProtobufWriter` is going out of scope. Does the current use allow this?

Reply:
Yes, ideally `release()` should be called when we no longer need the `ProtobufWriter` object. That's why I changed `m_buf` into a smart pointer: after `release()`, `m_buf` will be `nullptr` and will throw exception if you call `put_`.

Reply:
Exception is still a run time failure, I'd like it to fail earlier :)
ref-qualifying the member function would make the incorrect use difficult:
```suggestion
  std::vector<uint8_t> release() && { return std::move(m_buf); }
```
With this, the use should looks something like
```
auto const buff = [&](){
  ProtobufWriter pbw;

  // use protobuf writer
  ...

  return std::move(pbw).release();
}();
And with this I don't think there's need for the `unique_ptr` wrapper.

Reply:
What if somebody does this?
```
ProtobufWriter pbw;
...
// use pbw
...
auto buff = std::move(pbw).release();
// use buff
...
pbw.put_bytes(....); // <=== use pbw again?

```

Reply:
IMO it's an easy mistake to find, because an object is used after std::move. I don't think we can prevent such misuse in C++.

Reply:
I created a test with that:
```
#include <iostream>
#include <vector>

struct Buff {
    std::vector<int> buff;

    void put(int x) { buff.push_back(x); }

    auto release() && { return std::move(buff); }
};

int main() {
    Buff b;
    b.put(10);

    auto buff = std::move(b).release();
    std::cout << "buff: " << buff[0] << std::endl;

    b.put(1);
    std::cout << "PUT!!!" << std::endl;

    return 0;
}
```
The output:
```
Program returned: 0
buff: 10
PUT!!!
```

So we can use `pbw` after `std::move(pbw).release()` without any problem.

Reply:
This code invokes UB because of use after move. The move is explicit so IMO this is as much as we should do to prevent UB. 
I don't think we should design classes so that incorrect use causes exceptions instead of UB. Since we disagree, this is probably a good question for the rapids cpp channel.

Reply:
Using the version of `std::unique_ptr<std::vector>>`, the program crashes as I expected:
```
Program returned: 139
buff: 10
```

Note that there is no exception being thrown. If we want to throw, we need to check for `nullptr` in all the `put_` functions.

Reply:
> `std::move` is nothing more than a cast to an rvalue reference. In itself it doesn't actually move anything.

From my understanding, if you don't assign `std::move(...)` to anything then it will just be there.